File: python/cart_dmp/cart_coord_dmp/learnCartPrimitiveMultiOnLocalCoord.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
import copy
sys.path.append(os.path.join(os.path.dirname(__file__), '../../dmp_state/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../../dmp_param/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../../dmp_base/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../../dmp_discrete/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../../cart_dmp/cart_coord_dmp'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../../utilities/'))
from DMPTrajectory import *
from DMPState import *
from TauSystem import *
from DMPUnrollInitParams import *
from CanonicalSystemDiscrete import *
from CartesianCoordTransformer import *
from CartesianCoordDMP import *
from utilities import *

def learnCartPrimitiveMultiOnLocalCoord(cart_global_traj, train_data_dt,
                                        dmp_basis_funcs_size=25, canonical_order=2,
                                        ctraj_local_coordinate_frame_selection=GSUTANTO_LOCAL_COORD_FRAME,
                                        unroll_traj_length=-1,
                                        unroll_dt=None, 
                                        is_using_scaling=[False] * 3):
    # is_using_scaling=[False] * 3 # default is NOT using scaling on DMPs
    if (unroll_dt == None):
        unroll_dt = train_data_dt
    
    D = 3
    N_demo = len(cart_global_traj)
    
    tau_sys = TauSystem(MIN_TAU)
    canonical_sys_discr = CanonicalSystemDiscrete(tau_sys, canonical_order)
    cart_dmp = CartesianCoordDMP(dmp_basis_funcs_size, canonical_sys_discr, ctraj_local_coordinate_frame_selection)
    cart_dmp.setScalingUsage(is_using_scaling)
    
    [critical_states_learn, 
     W, mean_A_learn, mean_tau, 
     Ft, Fp, 
     G, cX, cV, 
     PSI] = cart_dmp.learnGetDefaultUnrollParams(cart_global_traj, 1.0/train_data_dt)
    
    if (unroll_traj_length == -1):
        unroll_tau = mean_tau
        unroll_traj_length = int(np.round(unroll_tau / dt)) + 1
    else:
        unroll_tau = 1.0 * unroll_dt * (unroll_traj_length - 1)
    
    ## Reproduce
    dmp_unroll_init_params = DMPUnrollInitParams(critical_states_learn, unroll_tau)
    
    cart_dmp.startWithUnrollParams(dmp_unroll_init_params)
    
    cart_coord_dmp_params = {}
    cart_coord_dmp_params['train_data_dt'] = train_data_dt
    cart_coord_dmp_params['model_size'] = dmp_basis_funcs_size
    cart_coord_dmp_params['canonical_order'] = canonical_order
    cart_coord_dmp_params['W'] = W
    cart_coord_dmp_params['A_learn'] = mean_A_learn
    cart_coord_dmp_params['mean_tau'] = mean_tau
    cart_coord_dmp_params['mean_start_global_position'] = copy.copy(cart_dmp.mean_start_global_position)
    cart_coord_dmp_params['mean_goal_global_position'] = copy.copy(cart_dmp.mean_goal_global_position)
    cart_coord_dmp_params['mean_start_local_position'] = copy.copy(cart_dmp.mean_start_local_position)
    cart_coord_dmp_params['mean_goal_local_position'] = copy.copy(cart_dmp.mean_goal_local_position)
    cart_coord_dmp_params['ctraj_local_coordinate_frame_selection'] = ctraj_local_coordinate_frame_selection
    cart_coord_dmp_params['T_local_to_global_H'] = copy.copy(cart_dmp.ctraj_hmg_transform_local_to_global_matrix)
    cart_coord_dmp_params['T_global_to_local_H'] = copy.copy(cart_dmp.ctraj_hmg_transform_global_to_local_matrix)
    cart_coord_dmp_params['unroll_dt'] = unroll_dt
    cart_coord_dmp_params['unroll_tau'] = unroll_tau
    
    list_cart_coord_dmpstate_unroll_fit_global = [None] * unroll_traj_length
    list_cart_coord_dmpstate_unroll_fit_local = [None] * unroll_traj_length
    list_forcing_term = [None] * unroll_traj_length
        
    for i in range(unroll_traj_length):
        time = 1.0 * (i*unroll_dt)
        
        [current_state_global, 
         current_state_local, 
         transform_sys_forcing_term, 
         transform_sys_coupling_term_acc, 
         transform_sys_coupling_term_vel, 
         func_approx_basis_function_vector] = cart_dmp.getNextState(unroll_dt, True)
        
        list_cart_coord_dmpstate_unroll_fit_global[i] = current_state_global
        list_cart_coord_dmpstate_unroll_fit_local[i] = current_state_local
        list_forcing_term[i] = transform_sys_forcing_term
    
    cart_coord_dmp_unroll_fit_global_traj = convertDMPStatesListIntoDMPTrajectory(list_cart_coord_dmpstate_unroll_fit_global)
    cart_coord_dmp_unroll_fit_local_traj = convertDMPStatesListIntoDMPTrajectory(list_cart_coord_dmpstate_unroll_fit_local)
    Ffit = np.hstack(list_forcing_term)
    
    return cart_coord_dmp_params, cart_coord_dmp_unroll_fit_global_traj, cart_coord_dmp_unroll_fit_local_traj, Ffit

File: matlab/cart_dmp/cart_coord_dmp/learnCartPrimitiveMultiOnLocalCoord.m
function [ varargout ] = learnCartPrimitiveMultiOnLocalCoord( varargin )
    % Author: Giovanni Sutanto
    % Date  : August 02, 2016
    
    cart_global_traj= varargin{1};
    train_data_dt  	= varargin{2};
    n_rfs           = varargin{3};
    c_order         = varargin{4};
    if (nargin > 4)
        ctraj_local_coordinate_frame_selection  = varargin{5};
    else
        ctraj_local_coordinate_frame_selection  = 1;    % default is gsutanto's ctraj_local_coordinate_frame_selection
    end
    if (nargin > 5)
        unroll_traj_length  = varargin{6};
    else
        unroll_traj_length  = -1;
    end
    if (nargin > 6)
        unroll_dt           = varargin{7};
    else
        unroll_dt           = train_data_dt;
    end
    if (nargin > 7)
        is_using_scaling    = varargin{8};
    else
        is_using_scaling    = zeros(1,3);   % default is NOT using scaling on DMPs
    end
    
    D               = 3;
    N_demo          = size(cart_global_traj,2);
    
    cart_local_traj     = cell(size(cart_global_traj));
    T_local_to_global_H = cell(1, N_demo);
    T_global_to_local_H = cell(1, N_demo);
    taus                = zeros(1, N_demo);
    dts                 = zeros(1, N_demo);
    mean_tau            = 0.0;
    mean_start_global   = zeros(3,1);
    mean_goal_global    = zeros(3,1);
    for i=1:N_demo
%         disp(['   Trajectory ', num2str(i), '/', num2str(N_demo)]);
        [cart_local_traj(1:3,i), T_local_to_global_H{1,i}, T_global_to_local_H{1,i}] = computeCartLocalTraj(cart_global_traj(1:3,i), 0, ctraj_local_coordinate_frame_selection);
        traj_length_i       = size(cart_local_traj{1,i},1);
        taus(1,i)           = ((traj_length_i-1)*train_data_dt);
        dts(1,i)            = train_data_dt;
        mean_tau            = mean_tau + taus(1,i);
        mean_start_global   = mean_start_global + (cart_global_traj{1,i}(1,:).');
        mean_goal_global    = mean_goal_global + (cart_global_traj{1,i}(end,:).');
%         keyboard;
    end
    mean_tau            = mean_tau / N_demo;
    mean_start_global   = mean_start_global / N_demo;
    mean_goal_global    = mean_goal_global / N_demo;
    
    w                   = zeros(n_rfs,D);
    dG                  = zeros(1,D);
        
    Ts                  = cell(1,N_demo);
    Tds                 = cell(1,N_demo);
    Tdds                = cell(1,N_demo);
    
    for d=1:D
        % learn primitive (per axis):
        dcp_franzi('init', d, n_rfs, num2str(d), c_order);
        
        for i=1:N_demo
            Ts{1,i}   	= cart_local_traj{1,i}(:,d);
            Tds{1,i}   	= cart_local_traj{2,i}(:,d);
            Tdds{1,i}  	= cart_local_traj{3,i}(:,d);
        end

        [w(:,d),~,~,~,~,~,~,~,~,...
         dG(:,d)]      	= dcp_franzi('batch_fit_multi',d,taus,dts,Ts,Tds,Tdds);
        
        if (is_using_scaling(1,d) == 0) % if NOT using scaling on this dimension's DMP:
            dG(:,d)     = 0;
        end
    end
    
%     cart_coord_dmp_params_basic.dt                  = train_data_dt;
    cart_coord_dmp_params_basic.dt                  = unroll_dt;
    cart_coord_dmp_params_basic.n_rfs               = n_rfs;
    cart_coord_dmp_params_basic.c_order             = c_order;
    cart_coord_dmp_params_basic.w                   = w;
    cart_coord_dmp_params_basic.dG                  = dG;
    unroll_cart_coord_params_basic.mean_tau         = mean_tau;
    unroll_cart_coord_params_basic.mean_start_global= mean_start_global;
    unroll_cart_coord_params_basic.mean_goal_global = mean_goal_global;
    unroll_cart_coord_params_basic.ctraj_local_coordinate_frame_selection   = ctraj_local_coordinate_frame_selection;
    
    if (nargout > 1)
        if (unroll_traj_length == -1)
            unroll_cart_coord_params_basic.mean_tau = mean_tau;
        else
            unroll_cart_coord_params_basic.mean_tau = unroll_dt * (unroll_traj_length - 1);
        end
        
        [cart_coord_dmp_params, ...
         cart_coord_dmp_unroll_fit_global_traj, ...
         cart_coord_dmp_unroll_fit_local_traj, ...
         Ffit]  = unrollCartPrimitiveOnLocalCoord( cart_coord_dmp_params_basic, ...
                                                   unroll_cart_coord_params_basic );
        
        varargout(1)    = {cart_coord_dmp_params};
        varargout(2)    = {cart_coord_dmp_unroll_fit_global_traj};
        varargout(3)    = {cart_coord_dmp_unroll_fit_local_traj};
        varargout(4)    = {Ffit};
        varargout(5)    = {mean_tau};
    else
        cart_coord_dmp_params   = completeCartCoordDMPParams( cart_coord_dmp_params_basic,...
                                                              unroll_cart_coord_params_basic );

        varargout(1)        	= {cart_coord_dmp_params};
    end
end